### PR TITLE
[CP-30] Added creditCardNumber pipe for viewing saved card numbers properly

### PR DESCRIPTION
**Pull request by AmGarera**
## Type of change

- [ ] Bug fix
- [x] New feature development
- [ ] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other

## Objective

Community Forum Posts
Feature request post: https://community.bitwarden.com/t/display-card-numbers-in-groups-to-make-them-easier-to-read/12042

GitHub contribution post: https://community.bitwarden.com/t/display-card-numbers-in-groups-to-make-them-easier-to-read/31541

Added the creditCardNumber pipe for viewing saved card numbers properly formatted. Supports all available brands selectable from the existing dropdown. PRs to the other repos will follow and be linked to this PR.
- https://github.com/bitwarden/browser/pull/2247
## Code changes

- **credit-card-number.pipe.ts** Added this file because it is a new pipe created to handle credit card formatting when viewing a saved card. referenced issues #427 
## Testing requirements
Test cards should be created for all available brands and their available known formats. I've done this locally so I can submit testing screenshots or whatever is deemed appropriate.

## Before you submit

- [x] I have checked for **linting** errors (`npm run lint`) (required)
- [ ] I have added **unit tests** where it makes sense to do so (encouraged but not required)
- [ ] This change requires a **documentation update** (notify the documentation team)
- [ ] This change has particular **deployment requirements** (notify the DevOps team)

### CODE REVIEW

**Comment by Hinton** [comment]
Hi @AmGarera,

Thanks for your contribution. This looks a bit difficult to maintain in the current format.

I think we need to look into creating a dictionary with the different lengths that can be more easily maintained, and reduse code duplication. Something similar to this:

```typescript
const numberFormats = {
  "visa": [
    [16, [4, 4, 4, 4]]
 ],
 "maestro": [
    [16, [4, 4, 4, 4]],
    [13, [4, 4, 5]],
    [14, [4, 6, 5]],
    [16, [4, 4, 4, 4, 3]],
  ],
}
```

If no card type is found, we should default the format to `4, 4, 4, 4`.

Using the lookup table we could then format the string accordingly.

**Comment by AmGarera** [comment]
@Hinton I've made changes based on your feedback. Please have another look when you have the time. Thanks.

File: angular/src/pipes/credit-card-number.pipe.ts [hunk 1]
```diff
@@ -0,0 +1,197 @@
+import { Pipe, PipeTransform } from '@angular/core';
+
+@Pipe({ name: 'creditCardNumber' })
+export class CreditCardNumberPipe implements PipeTransform {
+
+    numberFormats = {
```
Comment (Hinton) [inline angular/src/pipes/credit-card-number.pipe.ts:6]: Since we don't plan on modifying these lets move them outside the class as a `const`.

**Review by Hinton — CHANGES_REQUESTED** [review]
Apologies for the late review.

I believe we can clean up the code a bit further since it's a bit difficult to follow with all the loops. I've added a couple of suggestions which I believe will simplify the logic a fair bit while still retaining it's functionality.

Could you also please run `npm run prettier` which will make it a bit more readable.

Thanks!

**Comment by Hinton** [comment]
image
image

**Review by djsmith85 — APPROVED** [review]
Looks good.

@AmGarera Thank you very much for your contribution!

**Pull request merged**